hadoop/src/main/java/org/apache/carbondata/hadoop/api/DistributableDataMapFormat.java

```diff
@@ -0,0 +1,135 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.carbondata.hadoop.api;
+
+import java.io.IOException;
+import java.io.Serializable;
+import java.util.ArrayList;
+import java.util.Iterator;
+import java.util.List;
+
+import org.apache.carbondata.core.indexstore.Blocklet;
+import org.apache.carbondata.core.indexstore.DataMapDistributable;
+import org.apache.carbondata.core.indexstore.DataMapStoreManager;
+import org.apache.carbondata.core.indexstore.DataMapType;
+import org.apache.carbondata.core.indexstore.TableDataMap;
+import org.apache.carbondata.core.metadata.AbsoluteTableIdentifier;
+import org.apache.carbondata.core.scan.filter.resolver.FilterResolverIntf;
+import org.apache.carbondata.hadoop.util.ObjectSerializationUtil;
+
+import org.apache.hadoop.conf.Configuration;
+import org.apache.hadoop.mapreduce.InputSplit;
+import org.apache.hadoop.mapreduce.JobContext;
+import org.apache.hadoop.mapreduce.RecordReader;
+import org.apache.hadoop.mapreduce.TaskAttemptContext;
+import org.apache.hadoop.mapreduce.lib.input.FileInputFormat;
+
+/**
+ * Input format for datamaps, it makes the datamap pruning distributable.
+ */
+public class DistributableDataMapFormat extends FileInputFormat<Void, Blocklet> implements
+    Serializable {
+
+  private static final String FILTER_EXP = "mapreduce.input.distributed.datamap.filter";
+
+  private AbsoluteTableIdentifier identifier;
+
+  private String dataMapName;
+
+  private DataMapType dataMapType;
+
+  private List<String> validSegments;
+
+  public DistributableDataMapFormat(AbsoluteTableIdentifier identifier, String dataMapName,
+      DataMapType dataMapType, List<String> validSegments) {
+    this.identifier = identifier;
+    this.dataMapName = dataMapName;
+    this.dataMapType = dataMapType;
+    this.validSegments = validSegments;
+  }
+
+  public static void setFilterExp(Configuration configuration, FilterResolverIntf filterExp)
+      throws IOException {
+    if (filterExp != null) {
+      String string = ObjectSerializationUtil.convertObjectToString(filterExp);
+      configuration.set(FILTER_EXP, string);
+    }
+  }
+
+  private static FilterResolverIntf getFilterExp(Configuration configuration) throws IOException {
+    String filterString = configuration.get(FILTER_EXP);
+    if (filterString != null) {
+      Object toObject = ObjectSerializationUtil.convertStringToObject(filterString);
+      return (FilterResolverIntf) toObject;
+    }
+    return null;
+  }
+
+  @Override public List<InputSplit> getSplits(JobContext job) throws IOException {
+    TableDataMap dataMap =
+        DataMapStoreManager.getInstance().getDataMap(identifier, dataMapName, dataMapType);
+    List<DataMapDistributable> distributables = dataMap.toDistributable(validSegments);
```

Comment:
Can we first filter on the segment based on minmax ? If segments are more, even map reduce will take times